pkg/contract/chain.go

```diff
@@ -5,50 +5,336 @@ package contract
+	}
+}
+
+func GetBlockchainEndpoints(
+	app *application.Avalanche,
+	network models.Network,
+	chainSpec ChainSpec,
+	promptForRPCEndpoint bool,
+	promptForWSEndpoint bool,
+) (string, string, error) {
+	var (
+		rpcEndpoint string
+		wsEndpoint  string
+	)
+	switch {
+	case chainSpec.CChain:
+		rpcEndpoint = network.CChainEndpoint()
+		wsEndpoint = network.CChainWSEndpoint()
+	case network.Kind == models.Local || network.Kind == models.Devnet:
+		blockchainID, err := GetBlockchainID(app, network, chainSpec)
+		if err != nil {
+			return "", "", err
+		}
+		rpcEndpoint = network.BlockchainEndpoint(blockchainID.String())
+		wsEndpoint = network.BlockchainWSEndpoint(blockchainID.String())
+	case chainSpec.BlockchainName != "":
+		sc, err := app.LoadSidecar(chainSpec.BlockchainName)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to load sidecar: %w", err)
+		}
+		if sc.Networks[network.Name()].BlockchainID == ids.Empty {
+			return "", "", fmt.Errorf("blockchain has not been deployed to %s", network.Name())
+		}
+		if len(sc.Networks[network.Name()].RPCEndpoints) > 0 {
+			rpcEndpoint = sc.Networks[network.Name()].RPCEndpoints[0]
+		}
+		if len(sc.Networks[network.Name()].WSEndpoints) > 0 {
+			wsEndpoint = sc.Networks[network.Name()].WSEndpoints[0]
+		}
+	}
+	blockchainDesc, err := GetBlockchainDesc(chainSpec)
+	if err != nil {
+		return "", "", err
+	}
+	if rpcEndpoint == "" && promptForRPCEndpoint {
+		rpcEndpoint, err = app.Prompt.CaptureURL("Which is the RPC endpoint for "+blockchainDesc, false)
+		if err != nil {
+			return "", "", err
+		}
+	}
+	if wsEndpoint == "" && promptForWSEndpoint {
+		wsEndpoint, err = app.Prompt.CaptureURL("Which is the WS endpoint for "+blockchainDesc, false)
+		if err != nil {
+			return "", "", err
+		}
+	}
+	return rpcEndpoint, wsEndpoint, nil
 }
 
-func GetRPCURL(
+func GetBlockchainID(
 	app *application.Avalanche,
 	network models.Network,
-	subnetName string,
-	isCChain bool,
+	chainSpec ChainSpec,
+) (ids.ID, error) {
+	var blockchainID ids.ID
+	switch {
+	case chainSpec.BlockchainID != "":
+		var err error
+		blockchainID, err = ids.FromString(chainSpec.BlockchainID)
+		if err != nil {
+			// it should be an alias at this point
+			blockchainID, err = utils.GetChainID(network.Endpoint, chainSpec.BlockchainID)
+			if err != nil {
+				return ids.Empty, err
+			}
+		}
+	case chainSpec.CChain:
+		chainID, err := utils.GetChainID(network.Endpoint, "C")
+		if err != nil {
+			return ids.Empty, err
+		}
+		blockchainID = chainID
+	case chainSpec.BlockchainName != "":
+		sc, err := app.LoadSidecar(chainSpec.BlockchainName)
+		if err != nil {
+			return ids.Empty, fmt.Errorf("failed to load sidecar: %w", err)
+		}
+		if sc.Networks[network.Name()].BlockchainID == ids.Empty {
+			return ids.Empty, fmt.Errorf("blockchain has not been deployed to %s", network.Name())
+		}
+		blockchainID = sc.Networks[network.Name()].BlockchainID
```

Comment:
same here. we need to make sure map `sc.Networks`  has key `network.Name()`

Reply:
https://github.com/ava-labs/avalanche-cli/pull/2107#discussion_r1749295060